Lib/subprocess.py

```diff
@@ -1535,9 +1535,9 @@ def _communicate(self, input, endtime, orig_timeout):
                 self.stderr.close()
 
             # All data exchanged.  Translate lists into strings.
-            if stdout is not None:
+            if stdout:
                 stdout = stdout[0]
-            if stderr is not None:
+            if stderr:
                 stderr = stderr[0]
 
             return (stdout, stderr)
```

Comment:
There's an additional bug exposed by the above fix.  This line could now return ([], []) instead of (None, None).

Changing the above translation logic to

```
stdout = stdout[0] if stdout else None
stderr = stderr[0] if stdout else None
```

would work.  call this option [A[.

though nicer code all around would be to not use `_stdXXX_buff = []` at all but instead refactor `_readerthread` to take an attribute name to assign to instead of passing it a list to shove a single element in.

something similar to:

```
def _readerthread(self, fh, attr_name):
  try:
    setattr(self, attr_name, fh.read())
  finally:
    fh.close()
```

that larger refactoring makes sense in the context of 3.10.  call this option [B].

as a bugfix for 3.9 and earlier i'd stick with the modification of the stdout and stderr assignments.  The time when this case even happens looks to be only when the timeout has expired on the joins() above.  So what's broken is timeout logic on Windows during a timeout when collecting stdout and stderr.

lets go ahead with [A] in this PR and a followup PR can do [B].

Reply:
Updated to option A

Reply:
>  The time when this case even happens looks to be only when the timeout has expired on the joins() above. So what's broken is timeout logic on Windows during a timeout when collecting stdout and stderr.

If either thread is still alive, then a `TimeoutExpired` will be raised. The only time there is a problem is if one or more reader threads has failed, either by being forcibly terminated via `TerminateThread` (not exposed to Python code, except via ctypes) or if the `fh.read()` call fails, which is most likely due to another thread calling `CancelSynchronousIo` (also only exposed to Python code via ctypes).